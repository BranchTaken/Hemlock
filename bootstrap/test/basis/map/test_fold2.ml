open! Basis.Rudiments
open! Basis
open MapTest
open Map
open Format

let test () =
  printf "@[";
  let xpp_pair xppf (kv0_opt, kv1_opt) = begin
    fprintf xppf "(%a, %a)"
      (Option.xpp (xpp_kv Uns.xpp)) kv0_opt
      (Option.xpp (xpp_kv Uns.xpp)) kv1_opt
  end in
  let test ks0 ks1 = begin
    let map0 = of_klist ks0 in
    let map1 = of_klist ks1 in
    let pairs = fold2 ~init:[] ~f:(fun accum kv0_opt kv1_opt ->
      (kv0_opt, kv1_opt) :: accum
    ) map0 map1 in
    printf "fold2 %a %a -> %a@\n"
      (List.xpp Uns.xpp) ks0
      (List.xpp Uns.xpp) ks1
      (List.xpp xpp_pair) pairs
  end in
  let test_lists = [
    [];
    [0L];
    [0L; 1L];
    [0L; 1L; 2L];
    [0L; 1L; 66L];
    [0L; 1L; 66L; 91L];
    [42L; 420L];
    [42L; 420L; 421L];
    [42L; 420L; 4200L];
  ] in
  List.iteri test_lists ~f:(fun i ks0 ->
    List.iteri test_lists ~f:(fun j ks1 ->
      if i <= j then test ks0 ks1
    )
  )

let _ = test ()
