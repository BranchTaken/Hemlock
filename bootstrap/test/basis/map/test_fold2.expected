fold2 [] [] -> []
fold2 [] [0] -> [(None, Some (0, 0))]
fold2 [] [0; 1] -> [(None, Some (0, 0)); (None, Some (1, 100))]
fold2 [] [0; 1; 2] -> [(None, Some (0, 0)); (None, Some (2, 200)); (None, Some (1, 100))]
fold2 [] [0; 1; 66] -> [(None, Some (0, 0)); (None, Some (66, 6_600)); (None, Some (1, 100))]
fold2 [] [0; 1; 66; 91] -> [(None, Some (0, 0)); (None, Some (66, 6_600)); (None, Some (91, 9_100)); (None, Some (1, 100))]
fold2 [] [42; 420] -> [(None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [] [42; 420; 421] -> [(None, Some (421, 42_100)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [] [42; 420; 4_200] -> [(None, Some (4_200, 420_000)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0] [0] -> [(Some (0, 0), Some (0, 0))]
fold2 [0] [0; 1] -> [(Some (0, 0), Some (0, 0)); (None, Some (1, 100))]
fold2 [0] [0; 1; 2] -> [(Some (0, 0), Some (0, 0)); (None, Some (2, 200)); (None, Some (1, 100))]
fold2 [0] [0; 1; 66] -> [(Some (0, 0), Some (0, 0)); (None, Some (66, 6_600)); (None, Some (1, 100))]
fold2 [0] [0; 1; 66; 91] -> [(Some (0, 0), Some (0, 0)); (None, Some (66, 6_600)); (None, Some (91, 9_100)); (None, Some (1, 100))]
fold2 [0] [42; 420] -> [(Some (0, 0), None); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0] [42; 420; 421] -> [(Some (0, 0), None); (None, Some (421, 42_100)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0] [42; 420; 4_200] -> [(Some (0, 0), None); (None, Some (4_200, 420_000)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1] [0; 1] -> [(Some (0, 0), Some (0, 0)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1] [0; 1; 2] -> [(Some (0, 0), Some (0, 0)); (None, Some (2, 200)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1] [0; 1; 66] -> [(Some (0, 0), Some (0, 0)); (None, Some (66, 6_600)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1] [0; 1; 66; 91] -> [(Some (0, 0), Some (0, 0)); (None, Some (66, 6_600)); (None, Some (91, 9_100)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1] [42; 420] -> [(Some (0, 0), None); (Some (1, 100), None); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1] [42; 420; 421] -> [(Some (0, 0), None); (Some (1, 100), None); (None, Some (421, 42_100)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1] [42; 420; 4_200] -> [(Some (0, 0), None); (Some (1, 100), None); (None, Some (4_200, 420_000)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 2] [0; 1; 2] -> [(Some (0, 0), Some (0, 0)); (Some (2, 200), Some (2, 200)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1; 2] [0; 1; 66] -> [(Some (0, 0), Some (0, 0)); (Some (2, 200), None); (None, Some (66, 6_600)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1; 2] [0; 1; 66; 91] -> [(Some (0, 0), Some (0, 0)); (Some (2, 200), None); (None, Some (66, 6_600)); (None, Some (91, 9_100)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1; 2] [42; 420] -> [(Some (0, 0), None); (Some (2, 200), None); (Some (1, 100), None); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 2] [42; 420; 421] -> [(Some (0, 0), None); (Some (2, 200), None); (Some (1, 100), None); (None, Some (421, 42_100)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 2] [42; 420; 4_200] -> [(Some (0, 0), None); (Some (2, 200), None); (Some (1, 100), None); (None, Some (4_200, 420_000)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 66] [0; 1; 66] -> [(Some (0, 0), Some (0, 0)); (Some (66, 6_600), Some (66, 6_600)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1; 66] [0; 1; 66; 91] -> [(Some (0, 0), Some (0, 0)); (Some (66, 6_600), Some (66, 6_600)); (None, Some (91, 9_100)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1; 66] [42; 420] -> [(Some (0, 0), None); (Some (66, 6_600), None); (Some (1, 100), None); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 66] [42; 420; 421] -> [(Some (0, 0), None); (Some (66, 6_600), None); (Some (1, 100), None); (None, Some (421, 42_100)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 66] [42; 420; 4_200] -> [(Some (0, 0), None); (Some (66, 6_600), None); (Some (1, 100), None); (None, Some (4_200, 420_000)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 66; 91] [0; 1; 66; 91] -> [(Some (0, 0), Some (0, 0)); (Some (66, 6_600), Some (66, 6_600)); (Some (91, 9_100), Some (91, 9_100)); (Some (1, 100), Some (1, 100))]
fold2 [0; 1; 66; 91] [42; 420] -> [(Some (0, 0), None); (Some (66, 6_600), None); (Some (91, 9_100), None); (Some (1, 100), None); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 66; 91] [42; 420; 421] -> [(Some (0, 0), None); (Some (66, 6_600), None); (Some (91, 9_100), None); (Some (1, 100), None); (None, Some (421, 42_100)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [0; 1; 66; 91] [42; 420; 4_200] -> [(Some (0, 0), None); (Some (66, 6_600), None); (Some (91, 9_100), None); (Some (1, 100), None); (None, Some (4_200, 420_000)); (None, Some (420, 42_000)); (None, Some (42, 4_200))]
fold2 [42; 420] [42; 420] -> [(Some (420, 42_000), Some (420, 42_000)); (Some (42, 4_200), Some (42, 4_200))]
fold2 [42; 420] [42; 420; 421] -> [(None, Some (421, 42_100)); (Some (420, 42_000), Some (420, 42_000)); (Some (42, 4_200), Some (42, 4_200))]
fold2 [42; 420] [42; 420; 4_200] -> [(None, Some (4_200, 420_000)); (Some (420, 42_000), Some (420, 42_000)); (Some (42, 4_200), Some (42, 4_200))]
fold2 [42; 420; 421] [42; 420; 421] -> [(Some (421, 42_100), Some (421, 42_100)); (Some (420, 42_000), Some (420, 42_000)); (Some (42, 4_200), Some (42, 4_200))]
fold2 [42; 420; 421] [42; 420; 4_200] -> [(Some (421, 42_100), None); (None, Some (4_200, 420_000)); (Some (420, 42_000), Some (420, 42_000)); (Some (42, 4_200), Some (42, 4_200))]
fold2 [42; 420; 4_200] [42; 420; 4_200] -> [(Some (4_200, 420_000), Some (4_200, 420_000)); (Some (420, 42_000), Some (420, 42_000)); (Some (42, 4_200), Some (42, 4_200))]
